mm2src/coins/hd_wallet/mod.rs

```diff
@@ -0,0 +1,431 @@
+            accounts
+                .get_mut(&account_id)
+                .expect("getting an element should never fail due to the checks above")
+        }))
+    }
+
+    async fn get_accounts(&self) -> HDAccountsMap<Self::HDAccount> { self.get_accounts_mutex().lock().await.clone() }
+
+    async fn get_accounts_mut(&self) -> HDAccountsMut<'_, Self::HDAccount> { self.get_accounts_mutex().lock().await }
+
+    async fn remove_account_if_last(&self, account_id: u32) -> Option<Self::HDAccount> {
+        let mut x = self.get_accounts_mutex().lock().await;
+        // `BTreeMap::last_entry` is still unstable.
+        let (last_account_id, _) = x.iter().last()?;
+        if *last_account_id == account_id {
+            x.remove(&account_id)
+        } else {
+            None
+        }
+    }
+
+    async fn get_enabled_address(&self) -> Option<HDWalletAddress<Self>> {
+        let enabled_address = match self.enabled_address.clone() {
+            Some(id) => id,
+            None => return None,
+        };
+
+        let account = match self.get_account(enabled_address.account_id).await {
+            Some(account) => account,
+            None => return None,
+        };
+
+        let hd_address_id = HDAddressId {
+            chain: enabled_address.chain,
+            address_id: enabled_address.address_id,
+        };
+
+        let address = account
+            .derived_addresses()
+            .lock()
+            .await
+            .get(&hd_address_id)
+            .map(|addr| addr.address());
+
+        address
+    }
+}
+
+#[async_trait]
+impl<HDAccount> HDWalletStorageOps for HDWallet<HDAccount>
+where
+    HDAccount: HDAccountOps + Clone + Send + Sync,
+{
+    fn hd_wallet_storage(&self) -> &HDWalletCoinStorage { &self.hd_wallet_storage }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct HDAccountAddressId {
+    pub account_id: u32,
+    pub chain: Bip44Chain,
+    pub address_id: u32,
+}
+
+impl Default for HDAccountAddressId {
+    fn default() -> Self {
+        HDAccountAddressId {
+            account_id: 0,
+            chain: Bip44Chain::External,
+            address_id: 0,
+        }
+    }
+}
+
+impl From<StandardHDPath> for HDAccountAddressId {
+    fn from(der_path: StandardHDPath) -> Self {
+        HDAccountAddressId {
+            account_id: der_path.account_id(),
+            chain: der_path.chain(),
+            address_id: der_path.address_id(),
+        }
+    }
+}
+
+impl HDAccountAddressId {
+    pub fn to_derivation_path(
+        &self,
+        path_to_coin: &StandardHDPathToCoin,
+    ) -> Result<DerivationPath, MmError<Bip32Error>> {
+        let mut account_der_path = path_to_coin.to_derivation_path();
+        account_der_path.push(ChildNumber::new(self.account_id, true)?);
+        account_der_path.push(self.chain.to_child_number());
+        account_der_path.push(ChildNumber::new(self.address_id, false)?);
+        drop_mutability!(account_der_path);
```

Comment:
Is this really required for value that is immediately returned?

Reply:
It's not required, I like to add it in case a function is extended later. But for this specific case, the `to_derivation_path` function appears to be fixed and will never be subject to further extension.

Reply:
done